File: leet-code/110. Balanced Binary Tree/BalancedBinaryTree110_2.java
package com.example;

public class BalancedBinaryTree110_2 {
    public boolean isBalanced(TreeNode root) {
        return getHeight(root) != -1;
    }

    private int getHeight(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = getHeight(root.left);
        int right = getHeight(root.right);
        if (left == -1 || right == -1 || Math.abs(left - right) >= 1) {
            return -1;
        }
        return Math.max(left, right) + 1;
    }
}

File: leet-code/110. Balanced Binary Tree/BalancedBinaryTree110_2.kt
package com.example

import kotlin.math.abs
import kotlin.math.max

class BalancedBinaryTree110_2 {
    fun isBalanced(root: TreeNode?): Boolean {
        return dfs(root) != -1
    }

    private fun dfs(node: TreeNode?): Int {
        if (node == null) {
            return 0
        }
        val left = dfs(node.left)
        val right = dfs(node.right)
        if (left == -1 || right == -1 || abs(left - right) >= 2) {
            return -1
        }
        return max(left, right) + 1
    }
}
